docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+
+Cluster capacity consists of capacities of individual cluster nodes.
+Capacity covers cpu, memory, disk space and other resources.
+
+## Motivation and use cases
+
+Scheduler decides which node a pod gets scheduled on.
+The decision depends on many factors such as: available cpu, memory, disk, volumes, etc.
+As long as enough resources is available, pods are scheduled.
+On the other hand, when a pod can not be scheduled due to insufficient resources,
+admin is usually made aware too late.
+With knowledge of the current cluster capacity admin can be warned in advance.
+Resulting in addition of new nodes (horizontal scaling) or increase of resources of existing nodes (vertical scaling) before any pod becomes pending.
+
+With introduction of ``kubectl top`` command (https://github.com/kubernetes/kubernetes/pull/28844),
+an admin can see the current usage (cpu, memory, storage) of individual nodes.
+The command can answer question such as:
+
+* How many cpu/memory for a given node is used
+* How many pods are running and how many resources they consume
+
+Similar usage can be retrieved with `kubectl describe node` (look for `Capacity` and `Allocated resources`).
+`Pod` section shows individual pods and their resource consumption.
+
+Still, information about total amount of remaining capacity is not sufficient.
+As cluster capacity consists of capacities of individual nodes, the entire resource space is fragmented.
+At the same time as the Pod is the minimal schedulable unit, the cluster capacity is quantised.
+
+Thus, in order to get more precise estimation of remaining capacity,
+admin needs to be able to ask the cluster question "How many pods of given requirements can be scheduled".
+
+Depending on whether an administrator is interested in capacity of overall cluster or a subset of nodes,
+one needs to take into account:
+
+* Allocated memory for all pods
+* Allocated memory for each namespace
+* Allocated CPU for all pods
+* Allocated CPU for each namespace
+* Allocated PV attachments on the nodes
+
+To get better picture of how resources are computated take a look at [2].
+Resource usage of each pod is part of Pod status.
+If a pod is scheduled without sufficient resources, scheduling fails with `PodExceedsFreeCPU` or `PodExceedsFreeMemory`.
+
+### Pod requirements
+
+There are various requirements a pod can specify.
+As each pod consists of at least one container, administrator can set resource limits and requests [10] for each container.
+If set, container provides better information about consumed resources (e.g. memory, cpu).
+Each pod can specify a list of volumes.
+Some of the volumes can be claimed and provisioned with volume dynamic provisioner.
+Some pods are meant to be run on dedicated nodes.
+
+In general, requirements can include:
+
+* cpu
+* memory
+* required disk space (a.k.a disk bytes)
+* minimal disk IOPs
+* gpu cores
+* volumes
+* zones/nodes specified by labels, taints
```

Comment:
Also affinity on the pod and explicit affinity rules known only by the scheduler (in the scheduler config)